### PR TITLE
Add segmentation with the Chan-Vese active contours method

### LINKED ISSUE #335: [FEA] add cucim.skimage.segmentation.chan_vese
Once #318 is merged, we can add the `chan_vese` function for the Chan-Vese segmentation algorithm based on active contours. I will make a PR for this soon.

### PR DESCRIPTION
**Pull request by grlee77**
closes #335

This PR adds a function for segmentation based on active-contours. Test cases are directly from scikit-image, but the code itself was heavily refactored to make extensive use of kernel fusion via `cupy.fuse()`. I will add benchmark results here soon.

It should be reviewed after #318 is merged. The new commits only start from 502f41a

### CODE REVIEW

**Comment by grlee77** [comment]
This shows a huge speedup relative to scikit-image as image size increases (>200x):

shape | Duration (CPU) | Duration (GPU) | Acceleration
------|----------------|----------------|-------------
(256, 256) | 0.679 +/- 0.014 | 0.209 +/- 0.005 | 3.251
(512, 512) | 3.196 +/- 0.014 | 0.208 +/- 0.002 | 15.397
(1024, 1024) | 14.350 +/- 0.000 | 0.206 +/- 0.001 | 69.603
(2048, 2048) | 68.054 +/- 0.000 | 0.426 +/- 0.000 | 159.710
(3072, 3072) | 199.156 +/- 0.000 | 0.874 +/- 0.000 | 227.748

<details>
<summary>benchmark script</summary>

```Python
import cupy as cp
import matplotlib.pyplot as plt
from cupyx.profiler import benchmark
from skimage import data, segmentation

from cucim.skimage import img_as_float
from cucim.skimage.segmentation import chan_vese


_image = img_as_float(cp.array(data.camera()))

print("shape | Duration (CPU) | Duration (GPU) | Acceleration")
print("------|----------------|----------------|-------------")
for n in [0.5, 1, 2, 4, 6]:
    if n == 0.5:
        image = _image[128:-128, 128:-128]
    else:
        image = cp.tile(_image, (n, n))
    image_cpu = cp.asnumpy(image)
    kwargs = dict(mu=0.25, lambda1=1, lambda2=1, tol=1e-3, max_num_iter=200, dt=0.5, init_level_set="checkerboard", extended_output=True)
    perf_cpu = benchmark(segmentation.chan_vese, (image_cpu, ), kwargs=kwargs, n_warmup=0, n_repeat=10, max_duration=5)
    t_cpu = perf_cpu.gpu_times
    perf_gpu = benchmark(chan_vese, (image, ), kwargs=kwargs, n_warmup=4, n_repeat=100, max_duration=5)
    t_gpu = perf_gpu.gpu_times
    print(f"{image.shape} | {t_cpu.mean():0.3f} +/- {t_cpu.std():0.3f} | {t_gpu.mean():0.3f} +/- {t_gpu.std():0.3f} | {t_cpu.mean() / t_gpu.mean():0.3f}")
```

</details>

**Comment by jakirkham** [comment]
@grlee77, could you please take a look at the conflict here? Think it may be related to PR ( https://github.com/rapidsai/cucim/pull/318 )

**Comment by grlee77** [comment]
fixed the conflict

File: python/cucim/src/cucim/skimage/segmentation/_chan_vese.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+    if cp.max(image) != 0:
+        image = image / cp.max(image)
+
+    i = 0
+    if extended_output:
+        old_energy = _cv_energy(image, phi, mu, lambda1, lambda2)
+        energies = []
+    phivar = tol + 1
+
+    while(phivar > tol and i < max_num_iter):
```
Comment (jakirkham) [inline python/cucim/src/cucim/skimage/segmentation/_chan_vese.py:415]:
```suggestion
    while phivar > tol and i < max_num_iter:
```

**Review by jakirkham — COMMENTED** [review]
Thanks Greg! 🙏

Had a few comments below

File: python/cucim/src/cucim/skimage/segmentation/_chan_vese.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+    Csum += C2
+    Csum += C3
+    Csum += C4
+
+    return K, Csum
+
+
+@cp.fuse()
+def _fused_hphi_hinv(phi):
+    Hphi = (phi > 0).astype(phi.dtype)
```
[the remark this answers is not in the document]
Comment (grlee77) [reply python/cucim/src/cucim/skimage/segmentation/_chan_vese.py:79]: It works okay, but runs slower. I think we would still have to cast the output to float32 as well

File: python/cucim/src/cucim/skimage/segmentation/_chan_vese.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    grad2 = fx**2 + fy**2
+    K = (fxx * fy**2 - 2 * fxy * fx * fy + fyy * fx**2)
+    K /= (grad2 * cp.sqrt(grad2) + 1e-8)
+    return K
+
+
+def _cv_curvature(phi):
+    """Returns the 'curvature' of a level set 'phi'.
+    """
+    P = cp.pad(phi, 1, mode='edge')
```
[the remark this answers is not in the document]
Comment (grlee77) [reply python/cucim/src/cucim/skimage/segmentation/_chan_vese.py:26]: I'm not sure. Aside from the border it seems like it should be. Maybe we could do all but the border without padding and then handle the border explicitly. I did manage to reduce it from the [upstream implementation](https://github.com/scikit-image/scikit-image/blob/main/skimage/segmentation/_chan_vese.py#L22) where some of the same slices are taken multiple times.

General slicing within a CuPy fused kernel is not supported, but passing in the sliced arrays as done here works fine.

[the remark this answers is not in the document]
Comment (jakirkham) [reply python/cucim/src/cucim/skimage/segmentation/_chan_vese.py:26]: Yeah this is what I'm wondering. Maybe we can revisit after the release ( https://github.com/rapidsai/cucim/issues/372 ).

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
@gpucibot merge

**Comment by jakirkham** [comment]
Thanks Greg! 🙏

**Pull request merged**